presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java

```diff
@@ -0,0 +1,160 @@
+        this.step = AggregationNode.Step.partialInput(step);
+        this.expectedGroups = expectedGroups;
+        this.groupByPartialChannels = groupByPartialChannels.build();
+        this.hashChannel = hashChannel.isPresent() ? Optional.of(groupByTypes.size()) : hashChannel;
+        this.operatorContext = operatorContext;
+        this.spills = spills;
+        this.groupByTypes = groupByTypes;
+        this.systemMemoryContext = systemMemoryContext;
+        this.maxEntriesBeforeSpill = maxEntriesBeforeSpill;
+        this.memorySizeBeforeSpill = memorySizeBeforeSpill;
+        this.overwriteIntermediateChannelOffset = overwriteIntermediateChannelOffset;
+
+        rebuildHashAggregationBuilder();
+    }
+
+    @Override
+    public void processPage(Page page)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Iterator<Page> buildResult()
+    {
+        MergeSort mergeSort = new MergeSort(groupByTypes, hashAggregationBuilder.buildIntermediateTypes());
+        Iterator<Page> mergedSpilledPages = mergeSort.merge(spills);
+
+        return new Iterator<Page>() {
+            private Iterator<Page> resultPages = Collections.emptyIterator();
+
+            @Override
+            public boolean hasNext()
+            {
+                return mergedSpilledPages.hasNext() || resultPages.hasNext();
+            }
+
+            @Override
+            public Page next()
+            {
+                if (!resultPages.hasNext()) {
+                    rebuildHashAggregationBuilder();
+                    long memorySize = 0; // ensure that at least one merged page will be processed
+
+                    while (mergedSpilledPages.hasNext() && !shouldProduceOutput(memorySize)) {
+                        hashAggregationBuilder.processPage(mergedSpilledPages.next());
+                        memorySize = hashAggregationBuilder.getSizeInMemory();
+                        systemMemoryContext.setBytes(memorySize);
+                    }
+                    resultPages = hashAggregationBuilder.buildResult();
+                }
+
+                return resultPages.next();
+            }
+        };
+    }
+
+    @Override
+    public boolean checkFullAndUpdateMemory()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean isBusy()
+    {
+        return hashAggregationBuilder.isBusy();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        hashAggregationBuilder.close();
+    }
+
+    private boolean shouldProduceOutput(long memorySize)
+    {
+        return (maxEntriesBeforeSpill > 0 && hashAggregationBuilder.getGroupCount() > maxEntriesBeforeSpill)
+                || (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+    }
+
+    private void rebuildHashAggregationBuilder()
+    {
+        this.hashAggregationBuilder = new InMemoryHashAggregationBuilder(
```

Comment:
I think cleaner solution would be to pass `HashAggregationBuilder` as a constructor parameter and add `reset` method to `HashAggregationBuilder`.